src/Microsoft.VisualStudio.Extensibility.Testing.Xunit.Shared/Harness/VisualStudioInstanceKey.cs

```diff
@@ -53,22 +60,54 @@ public override int GetHashCode()
             hashCode = (hashCode * -1521134295) + Version.GetHashCode();
             hashCode = (hashCode * -1521134295) + EqualityComparer<string>.Default.GetHashCode(RootSuffix);
             hashCode = (hashCode * -1521134295) + MaxAttempts.GetHashCode();
+            hashCode = (hashCode * -1521134295) + EnvironmentVariables.Count.GetHashCode();
             return hashCode;
         }
 
         public string SerializeToString()
         {
-            return $"{(int)Version};{RootSuffix};{MaxAttempts}";
+            var builder = new StringBuilder();
+            builder.Append(EnvironmentVariables.Count);
+            foreach (var environmentVariable in EnvironmentVariables)
+            {
+                builder.Append(';');
+                builder.Append(Convert.ToBase64String(Encoding.UTF8.GetBytes(environmentVariable)));
+            }
+
+            return $"{(int)Version};{RootSuffix};{MaxAttempts};{builder}";
         }
 
         public static VisualStudioInstanceKey DeserializeFromString(string s)
         {
             var semicolon1 = s.IndexOf(';');
             var semicolon2 = s.IndexOf(';', semicolon1 + 1);
+            var semicolon3 = s.IndexOf(';', semicolon2 + 1);
+            var semicolon4 = s.IndexOf(';', semicolon3 + 1);
+            var environmentVariables = new string[0];
+
+            if (semicolon4 > 0)
+            {
+                var environmentVariableCount = int.Parse(s.Substring(semicolon3 + 1, semicolon4 - semicolon3 - 1));
+                Debug.Assert(environmentVariableCount > 0, $"Assertion failed: {nameof(environmentVariableCount)} > 0");
+
+                environmentVariables = new string[environmentVariableCount];
+                var currentStartSeparator = semicolon4;
+                for (var i = 0; i < environmentVariableCount; i++)
+                {
+                    var nextSeparator = s.IndexOf(';', currentStartSeparator + 1);
+                    var currentVariable = nextSeparator > 0
+                        ? s.Substring(currentStartSeparator + 1, nextSeparator - currentStartSeparator - 1)
+                        : s.Substring(currentStartSeparator + 1);
+                    environmentVariables[i] = Encoding.UTF8.GetString(Convert.FromBase64String(currentVariable));
+                    currentStartSeparator = nextSeparator;
```

Comment:
Do we have to do manual serialization?  Wondering if it would be simpler to use something like json here, or if there were restrictions that prevented that (e.g. don't want to load json serializing dlls)

Reply:
This needs to support a large number of different host environments, and not all of them support JSON.